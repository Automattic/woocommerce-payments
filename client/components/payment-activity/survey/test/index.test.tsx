/**
 * External dependencies
 */
import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import apiFetch from '@wordpress/api-fetch';

/**
 * Internal dependencies
 */
import Survey from '..';
import WcPayOverviewSurveyContext, {
	WcPayOverviewSurveyContextProvider,
} from '../context';

jest.mock( '@wordpress/api-fetch', () => jest.fn() );
jest.mock( '@wordpress/data', () => ( {
	useDispatch: jest.fn( () => ( { createErrorNotice: jest.fn() } ) ),
} ) );

describe( 'WcPayOverviewSurveyContextProvider', () => {
	afterEach( () => {
		jest.clearAllMocks();
	} );

	afterAll( () => {
		jest.restoreAllMocks();
	} );

	it( 'render the initial state', () => {
		const childrenMock = jest.fn().mockReturnValue( null );
		render(
			<WcPayOverviewSurveyContextProvider>
				<WcPayOverviewSurveyContext.Consumer>
					{ childrenMock }
				</WcPayOverviewSurveyContext.Consumer>
			</WcPayOverviewSurveyContextProvider>
		);

		expect( childrenMock ).toHaveBeenCalledWith( {
			setSurveyAnswers: expect.any( Function ),
			setSurveySubmitted: expect.any( Function ),
			surveyAnswers: {},
			surveySubmitted: false,
			responseStatus: 'resolved',
		} );
	} );

	it( 'should render survey emoji buttons and be clickable', () => {
		render(
			<WcPayOverviewSurveyContextProvider>
				<Survey />
			</WcPayOverviewSurveyContextProvider>
		);

		const surveyText = screen.getByText( 'Are these metrics helpful?' );
		expect( surveyText ).toBeInTheDocument();

		const buttons = screen.getAllByRole( 'button' );

		//show comments field
		fireEvent.click( buttons[ 0 ] );
		const commentLabel = screen.getByText(
			'Why do you feel that way? (optional)'
		);
		expect( commentLabel ).toBeInTheDocument();

		//show comments field
		fireEvent.click( buttons[ 1 ] );
		expect( commentLabel ).toBeInTheDocument();

		//show comments field
		fireEvent.click( buttons[ 2 ] );
		expect( commentLabel ).toBeInTheDocument();

		//hide comments field and submit survey
		fireEvent.click( buttons[ 3 ] );
		expect( commentLabel ).not.toBeInTheDocument();
		expect( apiFetch ).toHaveBeenCalledTimes( 1 );
	} );

	test( 'test survey initial display', () => {
		const { container } = render(
			<WcPayOverviewSurveyContextProvider>
				<Survey />
			</WcPayOverviewSurveyContextProvider>
		);
		expect( container ).toMatchSnapshot();
	} );

	test( 'test survey with comments textbox', () => {
		const { container } = render(
			<WcPayOverviewSurveyContextProvider>
				<Survey />
			</WcPayOverviewSurveyContextProvider>
		);

		const buttons = screen.getAllByRole( 'button' );
		fireEvent.click( buttons[ 1 ] );

		expect( container ).toMatchSnapshot();
	} );
} );
