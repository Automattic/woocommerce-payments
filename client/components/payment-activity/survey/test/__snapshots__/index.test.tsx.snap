// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WcPayOverviewSurveyContextProvider test survey initial display 1`] = `
<div>
  <div
    class="components-flex components-card__footer components-card-footer css-oc4v7j-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-small em57xhy0"
    data-wp-c16t="true"
    data-wp-component="CardFooter"
  >
    <div
      class="wcpay-payments-activity__survey"
    >
      <div
        class="survey_container"
      >
        Are those metrics helpful?
        <div
          class="survey_container__emoticons"
        >
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="very-unhappy"
              role="img"
            >
              😞
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="unhappy"
              role="img"
            >
              🫤
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="neutral"
              role="img"
            >
              😑
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="happy"
              role="img"
            >
              🙂
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="very-happy"
              role="img"
            >
              😍
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`WcPayOverviewSurveyContextProvider test survey with comments textbox 1`] = `
<div>
  <div
    class="components-flex components-card__footer components-card-footer css-oc4v7j-View-Flex-sx-Base-sx-Items-ItemsRow-Footer-borderRadius-borderColor-small em57xhy0"
    data-wp-c16t="true"
    data-wp-component="CardFooter"
  >
    <div
      class="wcpay-payments-activity__survey"
    >
      <div
        class="survey_container"
      >
        Are those metrics helpful?
        <div
          class="survey_container__emoticons"
        >
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="very-unhappy"
              role="img"
            >
              😞
            </span>
          </button>
          <button
            class="components-button has-icon selected"
            type="button"
          >
            <span
              aria-label="unhappy"
              role="img"
            >
              🫤
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="neutral"
              role="img"
            >
              😑
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="happy"
              role="img"
            >
              🙂
            </span>
          </button>
          <button
            class="components-button has-icon"
            type="button"
          >
            <span
              aria-label="very-happy"
              role="img"
            >
              😍
            </span>
          </button>
        </div>
      </div>
      <div
        class="close_container"
      >
        <button
          aria-label="Close dialog"
          class="components-button has-icon"
          type="button"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="28"
            viewBox="0 0 24 24"
            width="28"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 13.06l3.712 3.713 1.061-1.06L13.061 12l3.712-3.712-1.06-1.06L12 10.938 8.288 7.227l-1.061 1.06L10.939 12l-3.712 3.712 1.06 1.061L12 13.061z"
            />
          </svg>
        </button>
      </div>
      <hr />
      <div
        class="comment_container"
      >
        <div
          class="components-base-control css-wdf2ti-Wrapper ej5x27r4"
        >
          <div
            class="components-base-control__field css-10urnh1-StyledField-deprecatedMarginField ej5x27r3"
          >
            <label
              class="components-base-control__label css-eweeby-StyledLabel-labelStyles ej5x27r2"
              for="inspector-textarea-control-1"
            >
              Why do you feel that way? (optional)
            </label>
            <textarea
              class="components-textarea-control__input css-zika3g-StyledTextarea-inputStyleNeutral-inputStyleFocus-inputControl ebk7yr50"
              id="inspector-textarea-control-1"
              rows="4"
            />
          </div>
        </div>
        <p
          class="comment_container__disclaimer"
        >
          Your feedback will be only be shared with WooCommerce and treated pursuant to our 
          <a
            href="https://automattic.com/privacy/"
            rel="noreferrer"
            target="_blank"
          >
            privacy policy
          </a>
          .
        </p>
      </div>
      <div
        class="wcpay-confirmation-modal__footer"
      >
        <button
          class="components-button is-tertiary"
          type="button"
        >
          Cancel
        </button>
        <button
          class="components-button is-primary"
          type="button"
        >
          Send
        </button>
      </div>
    </div>
  </div>
</div>
`;
